### PR TITLE
A very basic working example to learn samples

### PR DESCRIPTION
**Pull request by chaithyagr**
This had to be setup after discussing with @alineyyy , where we found some issues due to which it was harder to instance the operator inside a torch module. @paquiteau , maybe a quick review of the code to ensure we agree on style (we can meet if required). @alineyyy please update the learn samples codes to make it cleaner. Please also add a simple projection step to ensure we sample the center of k-space.

### CODE REVIEW

File: examples/example_learn_samples.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+class Model(torch.nn.Module):
+    def __init__(self):
+        super(Model, self).__init__()
+        self.operator = get_operator("gpunufft", wrt_data=True, wrt_traj=True)(
+            initialize_2D_radial(10, 192).reshape(-1, 2).astype(np.float32),
+            shape=(256, 256),
+            density=True,
+            squeeze_dims=False,
+        )
+        self.trajectory = self.operator.samples_torch
```
Comment (chaithyagr) [inline examples/example_learn_samples.py:36]: This is just done for simplicity

File: src/mrinufft/_utils.py [hunk 2]
```diff
@@ -86,7 +86,7 @@ def proper_trajectory(trajectory, normalize="pi"):
         raise ValueError(
             "trajectory should be array_like, with the last dimension being coordinates"
         ) from e
-
+
```
Comment (chaithyagr) [inline src/mrinufft/_utils.py:89]:
```suggestion
```

File: src/mrinufft/operators/autodiff.py [hunk 3]
```diff
@@ -135,15 +137,25 @@ def __init__(self, nufft_op, wrt_data=True, wrt_traj=False):
...
-        return _NUFFT_ADJOP.apply(kspace, self.samples, self.nufft_op)
+        return _NUFFT_ADJOP.apply(kspace, self.samples_torch, self.nufft_op)
 
     def __getattr__(self, name):
         """Get the attribute from the root operator."""
-        return getattr(self.nufft_op, name)
+        if hasattr(self.nufft_op, name):
+            if name == 'samples':
+                return self.samples_torch
+            return getattr(self.nufft_op, name)
```
Comment (paquiteau) [inline src/mrinufft/operators/autodiff.py:159]: Don't make the exception here, just add a property. Also this will break if you don't have it initialized. 

```python
@property
def samples(self):
    try:
        return self._samples_torch 
    except AttributeError:
        return self.nufft_op.samples

def __getattr__(self, name):
    return getattr(self.nufft_op, name)
```

File: src/mrinufft/operators/autodiff.py [hunk 4]
```diff
@@ -135,15 +137,25 @@ def __init__(self, nufft_op, wrt_data=True, wrt_traj=False):
...
+        else:
+            return super().__getattr__(name)
```
[the remark this answers is not in the document]
Comment (paquiteau) [reply src/mrinufft/operators/autodiff.py:161]: see above

File: tests/operators/test_autodiff.py [hunk 5]
```diff
@@ -75,15 +75,14 @@ def test_adjoint_and_grad(operator, interface):
...
         pytest.skip("GPU not supported for finufft backend")
 
-    if torch.is_tensor(operator.samples):
-        operator.samples = operator.samples.cpu().detach().numpy()
-
-    operator.samples = to_interface(operator.samples, interface=interface)
+    if "gpu" in interface:
+        operator.to("cuda")
+    else:
+        operator.cpu()
```
Comment (paquiteau) [inline tests/operators/test_autodiff.py:81]: what does this mean ? operator is not a tensor here.

File: examples/example_learn_samples.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+from mrinufft import check_backend, get_density, get_operator
+from mrinufft.trajectories import initialize_2D_radial
+from mrinufft.trajectories.display import display_2D_trajectory
+
+# %%
+# Create sample data
+# ------------------
+
+class Model(torch.nn.Module):
+    def __init__(self):
```
Comment (paquiteau) [inline examples/example_learn_samples.py:28]: Why do you need the extra module layer ?

**Review by paquiteau — CHANGES_REQUESTED** [review]
A good start, adding an example for the autodiff is clearly required ! However I would recommend to add more detail in the example to explains what is happening.

File: examples/example_learn_samples.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+Density Compensation Routines
+=============================
+
+Examples of differents density compensation methods.
+
+Density compensation depends on the sampling trajectory,and is apply before the
+adjoint operation to act as preconditioner, and should make the lipschitz constant
+of the operator roughly equal to 1.
+
+"""
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply examples/example_learn_samples.py:13]: AHh no no, its just a temporary file. Let me mark this as draft

File: src/mrinufft/operators/autodiff.py [hunk 8]
```diff
@@ -135,15 +137,25 @@ def __init__(self, nufft_op, wrt_data=True, wrt_traj=False):
         if wrt_traj and self.nufft_op.backend in ["finufft", "cufinufft"]:
             self.nufft_op.raw_op._make_plan_grad()
         self.nufft_op._grad_wrt_data = wrt_data
-
+        if wrt_traj:
+            self.samples_torch = torch.nn.Parameter(
+                data=torch.Tensor(self.nufft_op.samples),
+                requires_grad=True,
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply src/mrinufft/operators/autodiff.py:143]: But samples will route to samples_torch. But I understand your point, lemme do it. Its just that I am wondering if torch can handle the routing and properties right for autodiff.

File: src/mrinufft/operators/autodiff.py (hunk 3 above)
Comment (chaithyagr) [reply src/mrinufft/operators/autodiff.py:159]: Ahhh good point, but then if we are in autodiff module with grad_wrt_traj, we SHOULD have it initialized. 
But I dont know if it misses any corner cases (perhaps if we want grad_wrt_data and request `samples`

File: tests/operators/test_autodiff.py (hunk 5 above)
Comment (chaithyagr) [reply tests/operators/test_autodiff.py:81]: `operator` is a `torch.nn.module` here. This means all the underlying parameters / weights will be on GPU or CPU. It is very "torch-like" code, and this is good for us

File: examples/example_learn_samples.py (hunk 6 above)
Comment (chaithyagr) [reply examples/example_learn_samples.py:28]: I think and expect this is how users will use our operator. The `test` handles the case when they dont instance in a module and the example handles it being instanced in a module. Note that this was failing earlier, and hence we have this PR to begin with. Eventually we might want the model to dynamically update the shape of the trajectory (during upsampling steps for example.

**Comment by chaithyagr** [comment]
Closing as #167 is now here!
